src/acceptor.rs

```diff
@@ -0,0 +1,535 @@
+    #[no_mangle]
+    pub extern "C" fn rustls_acceptor_new(
+        acceptor_out: *mut *mut rustls_acceptor,
+    ) -> rustls_result {
+        ffi_panic_boundary! {
+            match Acceptor::new() {
+                Ok(acceptor) => {
+                    BoxCastPtr::set_mut_ptr(acceptor_out, acceptor);
+                    rustls_result::Ok
+                },
+                Err(e) => map_error(e),
+            }
+        }
+    }
+
+    /// Free a rustls_acceptor. Calling with NULL is fine.
+    /// Must not be called twice with the same value.
+    #[no_mangle]
+    pub extern "C" fn rustls_acceptor_free(acceptor: *mut rustls_acceptor) {
+        ffi_panic_boundary! {
+            BoxCastPtr::to_box(acceptor);
+        }
+    }
+
+    /// Check if this rustls_acceptor wants additional TLS bytes read into it.
+    /// If this returns true, you should call rustls_acceptor_read_tls().
+    #[no_mangle]
+    pub extern "C" fn rustls_acceptor_wants_read(acceptor: *const rustls_acceptor) -> bool {
+        ffi_panic_boundary! {
+            let acceptor = try_ref_from_ptr!(acceptor);
+            acceptor.wants_read()
+        }
+    }
+
+    /// Read some TLS bytes from the network into internal buffers. The actual network
+    /// I/O is performed by `callback`, which you provide. Rustls will invoke your
+    /// callback with a suitable buffer to store the read bytes into. You don't have
+    /// to fill it up, just fill with as many bytes as you get in one syscall.
+    /// The `userdata` parameter is passed through directly to `callback`. Note that
+    /// this is distinct from the `userdata` parameter set with
+    /// `rustls_connection_set_userdata`.
+    ///
+    /// Returns 0 for success, or an errno value on error. Passes through return values
+    /// from callback. See rustls_read_callback for more details.
+    /// If this returns success, you should call rustls_acceptor_accept().
+    #[no_mangle]
+    pub extern "C" fn rustls_acceptor_read_tls(
+        acceptor: *mut rustls_acceptor,
+        callback: rustls_read_callback,
+        userdata: *mut c_void,
+        out_n: *mut size_t,
+    ) -> rustls_io_result {
+        ffi_panic_boundary! {
+            let acceptor: &mut Acceptor = try_mut_from_ptr!(acceptor);
+            if out_n.is_null() {
+                return rustls_io_result(EINVAL);
+            }
+            let callback: ReadCallback = try_callback!(callback);
+
+            let mut reader = CallbackReader { callback, userdata };
+
+            let n_read: usize = match acceptor.read_tls(&mut reader) {
+                Ok(n) => n,
+                Err(e) => return rustls_io_result(e.raw_os_error().unwrap_or(EIO)),
+            };
+            unsafe {
+                *out_n = n_read;
+            }
+
+            rustls_io_result(0)
+        }
+    }
+
+    /// Parse all TLS bytes read so far. Returns:
+    ///
+    /// - RUSTLS_RESULT_NOT_READY: a full ClientHello has not yet been read.
```

Comment:
I think this could use some more explanation on the state of `acceptor` after calling this. Should the caller free `acceptor` after getting `RUSTLS_RESULT_OK` or another `RUSTLS_RESULT_NOT_READY` return value?

Reply:
Updated. Let me know if you think it's clearer.

Reply:
Yes, this is pretty okay now.